### ISSUE TITLE
Adapter does not seem to respect any TestCategory filtering

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit3-vs-adapter#452]
At risk of being overly broad, it's still true that I haven't come up with a way to pass a TestCategory filter to it and actually have the filter do anything (except cause explicit tests to indiscriminately run, but that's a separate issue). However I wonder if these specific examples aren't all the same issue:

* https://github.com/nunit/nunit3-vs-adapter/issues/425 Run only specific tests when using dotnet test?
* https://github.com/nunit/nunit3-vs-adapter/issues/438 vstest.console's `--TestCaseFilter` is ignored when listing tests
* Maybe? https://github.com/nunit/nunit3-vs-adapter/issues/241 Nunit test runner picking up ignored tests (ish) in tfs

Add to that the inability I have of [getting any category filter to work](https://github.com/nunit/nunit3-vs-adapter/issues/296#issuecomment-365760488) which is blocking the PR of the Mono.Cecil removal.

-----

VSTest.Console.exe is getting here:

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L65-L73

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L96

**Why are we using TestFilter.Empty here?**

Not here:

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L109-L115

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L141

**Comment by jnm2** [comment]
Here is the commit that introduced `TestFilter.Empty`: [Refactor in preparation for issue 47](https://github.com/nunit/nunit3-vs-adapter/commit/86a559497b5e99c8bafe8f7d7e71c03976e0a97e?diff=split#diff-b6035503adb3f6fc2b0cf4ff6598c1e6R68)

https://github.com/nunit/nunit3-vs-adapter/issues/47 is "Test Adapter runs explicit tests when TFS TestCaseFilter is used."

**Comment by CharliePoole** [comment]
@jnm2 Take the following with a grain of salt as I have not worked on the adapter since the .NET Core build was added and @rprouse or @OsirisTerje may have added something. But background info can be helpful sometimes if you take it for what it is.

The adapter has never had a way to accept NUnit filters from outside. It does use them internally. For example, when VS passes a list of test names, it constructs a name filter for those names. When a VS-style filter is passed, it is used to select a set of tests and then a name filter is constructed for those tests.

Way back when, probably on one of those issues you cited, I suggested two ways we could solve the "Explicit Problem"... (1) Accepting NUnit-style filter expressions in the .runsettings file or (2) creating a parser for the VS filter expressions, which would output a correct NUnit-style filter. Each of those represented major amounts of work and since I already knew I would be stepping down I wanted other team members to select the strategy. AFAIR, there was no feedback so neither option was ever implemented. The adapter - AFAIK - still does not accept any NUnit filter from the command-line or the .runsettings file. 

**********

Regarding your question about `TestFilter.Empty`, it is used in the first method because that's the method that is called in response to `Run All`, whereas the other is called in response to running selected tests.

However, the first method is also called when a filter is provided. That is taken care of in the RunAssembly method, which replaces the provided NUnit filter with a new one that is constructed from the one used. Probably not the cleanest design, but at the time, we didn't actually know which methods might be called by VS, so I put it in a common method. (Come to think of it, we still don't know that a lot of the time.) The provided filter is called a "TFS Filter" because that was originally the only way to provide us with a filter. If you debug down into RunAssembly, you'll find out if the filter you provide on the command-line is seen or not. That's where the problem is no doubt arising.

**Comment by jnm2** [comment]
Thanks Charlie! That's really helpful.

Hopefully we can end up with something that works without haveing to make assumptions about when and why methods are called by VS!

**Comment by jnm2** [comment]
@CharliePoole The thing that convinced me that `TestFilter.Empty` was never replaced with an actual filter was this:

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L204-L213

It prints `Running all tests in assembly.dll`.

**Comment by jnm2** [comment]
So the problem seems to be that a trait named "Category" does not show up in the TestCase.Traits collection:

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/TFSTestFilter.cs#L148-L157

Though it does show up in the TestCase.Properties collection as a property with the label `"TestCategory"`, ID `"NUnit.TestCategory"`, and single value `"Navigation"`.


I've never seen this code before, so I'm hoping for pointers. :o)

**Comment by jnm2** [comment]
Here's the contradiction. When **filtering**, the only recognized 'VS TestCase property' labels are `"FullyQualifiedName"` and `"Name"`. The only recognized 'VS TestCase trait' names are `"Priority"` and `"TestCategory"`. Thus, category is a **trait but not a property**.

But when **constructing** TestCases, the adapter decides that 'NUnit properties' become 'VS TestCase traits' *unless* the 'NUnit property' name is `"Category"` whereupon they go into a 'VS TestCase property' with the label `"TestCategory"`. Thus, category is a **property but not a trait**:

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/CategoryList.cs#L42-L47

https://github.com/nunit/nunit3-vs-adapter/blob/126a67701468055e9b43f1647f862a15440c0e85/src/NUnitTestAdapter/CategoryList.cs#L78-L84

So **one of the two** needs to change. Which one?

**Comment by jnm2** [comment]
Oh hey, this was a recent change! I should watch the repo closer just to maintain familiarity with this stuff.  https://github.com/nunit/nunit3-vs-adapter/pull/446

**Comment by jnm2** [comment]
Oh hey, this was a recent change! I should watch the repo closer just to maintain familiarity with this stuff.  https://github.com/nunit/nunit3-vs-adapter/pull/446

Reading the PR and issue, it looks like the construction phase should stay as-is. Therefore I'll correct the filtering stage after adding a failing test to cover this fix.

**Comment by CharliePoole** [comment]
I added the traits code somwhere around 2012, having gotten it from the Microsoft developers. IIRC, there was no such thing as a TestCategory back then, there were only traits. We created traits from each NUnit property and they worked fine in the IDE. Since filters were only used in TFS and I didn't use TFS, they got very little attention from me.

The upshot, is that there's very little I can give you in terms of pointers. Maybe @OsirisTerje can help. He should decide which way to go with this anyway.
